File: sources/lua/06_char_count/main.lua
-- task 06 char_count — expected output: 10000000
-- build: none (interpreted)    run: lua main.lua (PUC Lua 5.4)    also runs on LuaJIT: luajit main.lua
-- The 100 MB text is built once with string.rep, then scanned one character at a time.

local text = string.rep("abcdefghij", 10000000)

local count = 0
for i = 1, #text do
    local ch = string.byte(text, i)
    if ch == 97 then          -- 'a': skip
    elseif ch == 101 then     -- 'e': skip
    elseif ch == 104 then     -- 'h': count
        count = count + 1
    end
end

print(count)

File: sources/java/06_char_count/Main.java
// task 06 char_count — expected output: 10000000
// build: javac Main.java    run: java Main
// build (graalvm native-image): native-image -O2 Main    run: ./Main
// The 100 MB text is built once with String.repeat, then scanned one char at a time.

public class Main {
    public static void main(String[] args) {
        String text = "abcdefghij".repeat(10000000);
        long count = 0;
        for (int i = 0; i < text.length(); i++) {
            if (text.charAt(i) == 'h') {
                count += 1;
            }
        }
        System.out.println(count);
    }
}
